### DISCUSSION TITLE
Retry to connect to database

### DISCUSSION THREAD

**Discussion by leonwind** [dropwizard/dropwizard#4774] [General]
If I run the following code to retry to connect to my database the exception never gets caught since the run method catches the exception already by itself.

```java
public static void main(String[] args) throws Exception {
    while (true) {
      try {
        new Application().run(args);
        return;
      } catch (SQLException ex) {
        System.out.println("Retry connecting to database...");
        continue;
      }
    }
  }
```

Is there any workaround for this without editing the code of the run method so I can catch the exception by myself?

Thanks,
Leon

**Comment by joschi** [comment]
What exactly are you trying to achieve with your try-catch block?

**Comment by leonwind** [comment]
I am trying to wait until my database is ready. Currently there is just a SQLException which gets thrown and then the application is terminating and I want that it keeps trying until the database is ready

**Comment by joschi** [comment]
@leonwind Could you share a more details about your application so that we can weigh in what approach to take?

**Comment by leonwind** [comment]
I am running my dropwizard service and my database in two docker containers. The problem is that the service finishes building before the database and then crashes since it is not able to connect with it. I can add the attribute "depends_on" in my dockerfile but this only ensures that the service starts after the database. But if the database needs to set up the schema it still is not finished before the service.
Thus I would like to check in my service if the database is up and running and if it is not yet running, I want to wait until it is running and I do not want the service to crash. However, I am unable to catch the exception in the service.

**Reply by maffe** [reply]
With the [long syntax](https://docs.docker.com/compose/compose-file/#depends_on) in Docker Compose you can specify the database must be healthy (reachable) before the other container is started.

**Reply by leonwind** [reply]
Ohh awesome!!! I currently just added a time sleep in my service start script but this is definitely the better solution!
Thanks for also answering two years later :)

**Comment by joschi** [comment]
@leonwind I'm unable to reproduce your issue with this simple Dropwizard project:
https://github.com/joschi/dropwizard-issue-3550

Even if there is no database available, it only throws an exception when trying to obtain the `Connection` but keeps working otherwise.

**Comment by iamabs2001** [comment]
@leonwind you can created your own exception as per your requirement or follow asyn
